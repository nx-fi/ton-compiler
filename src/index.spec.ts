/**
 * Copyright (c) Whales Corp. 
 * All Rights Reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import { compileContract } from ".";
import path from 'path';


describe('ton-compiler', () => {
  it('should crash if sources are not under working dir', async () => {
    await expect(compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'legacy', workdir: path.resolve(__dirname, 'engines') })).rejects.toThrowError();
  });
  it.skip('should compile source with legacy compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'legacy' });
    expect(compiled).toMatchSnapshot();
    compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'global.fc')], version: 'legacy' });
    expect(compiled).toMatchSnapshot();
  });
  it.skip('should compile with errors with legacy compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test_bug.fc')], version: 'legacy' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile source with v2022.10 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'v2022.10' });
    expect(compiled).toMatchSnapshot();
    compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'global.fc')], version: 'v2022.10' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile with errors with v2022.10 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test_bug.fc')], version: 'v2022.10' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile source with v2022.12 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'v2022.12' });
    expect(compiled).toMatchSnapshot();
    compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'global.fc')], version: 'v2022.12' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile with errors with v2022.12 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test_bug.fc')], version: 'v2022.12' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile source with v2023.01 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'v2023.01' });
    expect(compiled).toMatchSnapshot();
    compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'global.fc')], version: 'v2023.01' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile with errors with v2023.01 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test_bug.fc')], version: 'v2023.01' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile source with v2024.01 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test.fc')], version: 'v2024.01' });
    expect(compiled).toMatchSnapshot();
    compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'global.fc')], version: 'v2024.01' });
    expect(compiled).toMatchSnapshot();
  });
  it('should compile with errors with v2024.01 compiler', async () => {
    let compiled = await compileContract({ files: [path.resolve(__dirname, 'tests', 'test_bug.fc')], version: 'v2024.01' });
    expect(compiled).toMatchSnapshot();
  });
});