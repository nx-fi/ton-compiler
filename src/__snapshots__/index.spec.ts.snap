// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ton-compiler should compile source with legacy compiler 1`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC recv_internal
  DECLPROC recv_external
  recv_internal PROC:<{
    //  in_msg
    DROP	// 
  }>
  recv_external PROC:<{
    //  in_msg
    9 PUSHPOW2	//  in_msg _3=512
    LDSLICEX	//  signature in_msg
    DUP	//  signature in_msg cs
    32 LDU	//  signature in_msg msg_seqno cs
    c4 PUSH	//  signature in_msg msg_seqno cs _11
    CTOS	//  signature in_msg msg_seqno cs cs2
    32 LDU	//  signature in_msg msg_seqno cs stored_seqno cs2
    256 LDU	//  signature in_msg msg_seqno cs stored_seqno public_key cs2
    ENDS
    s3 s1 XCPU	//  signature in_msg public_key cs stored_seqno msg_seqno stored_seqno
    EQUAL	//  signature in_msg public_key cs stored_seqno _23
    33 THROWIFNOT
    s0 s3 XCHG	//  signature stored_seqno public_key cs in_msg
    HASHSU	//  signature stored_seqno public_key cs _26
    s0 s4 s2 XC2PU	//  cs stored_seqno public_key _26 signature public_key
    CHKSIGNU	//  cs stored_seqno public_key _27
    34 THROWIFNOT
    ACCEPT
    s2 s0 XCPU	//  public_key stored_seqno cs cs
    SREFS	//  public_key stored_seqno cs _32
    IF:<{	//  public_key stored_seqno cs
      8 LDU	//  public_key stored_seqno mode cs
      LDREF	//  public_key stored_seqno mode _37 cs
      s0 s2 XCHG	//  public_key stored_seqno cs _37 mode
      SENDRAWMSG
    }>	//  public_key stored_seqno cs
    ENDS
    INC	//  public_key _42
    NEWC	//  public_key _42 _43
    32 STU	//  public_key _45
    256 STU	//  _47
    ENDC	//  _48
    c4 POP
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      79,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      1,
      32,
      2,
      3,
      0,
      4,
      210,
      48,
      0,
      110,
      242,
      131,
      8,
      215,
      24,
      32,
      211,
      31,
      237,
      68,
      208,
      211,
      31,
      211,
      255,
      209,
      81,
      49,
      186,
      242,
      161,
      3,
      249,
      1,
      84,
      16,
      66,
      249,
      16,
      242,
      162,
      248,
      0,
      81,
      32,
      215,
      74,
      150,
      211,
      7,
      212,
      2,
      251,
      0,
      222,
      209,
      164,
      200,
      203,
      31,
      203,
      255,
      201,
      237,
      84,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with legacy compiler 2`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC do_something
  DECLPROC do_another
  DECLPROC get_first
  DECLPROC get_second
  DECLPROC do_mutate
  DECLPROC main
  DECLGLOBVAR ctx
  do_something PROC:<{
    //  a
    DUP	//  a a
  }>
  do_another PROC:<{
    //  a b
    ADD	//  _2
  }>
  get_first PROC:<{
    //  a b
    DROP	//  a
  }>
  get_second PROC:<{
    //  a b
    NIP	//  b
  }>
  do_mutate PROC:<{
    //  a b c
    s2 s1 PUXC	//  a c a b
    ADD	//  a c _3
    -ROT	//  _3 a c
    ADD	//  _3 _4
  }>
  main PROC:<{
    // 
    1 PUSHINT	//  _1=1
    do_something CALLDICT	//  x x
    PAIR
    ctx SETGLOB
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      32,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      2,
      210,
      2,
      3,
      0,
      15,
      28,
      124,
      0,
      91,
      192,
      190,
      24,
      96,
      0,
      3,
      8,
      32,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2022.10 compiler 1`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC recv_internal
  DECLPROC recv_external
  recv_internal PROC:<{
    DROP
  }>
  recv_external PROC:<{
    9 PUSHPOW2
    LDSLICEX
    DUP
    32 LDU
    c4 PUSH
    CTOS
    32 LDU
    256 LDU
    ENDS
    s3 s1 XCPU
    EQUAL
    33 THROWIFNOT
    s0 s3 XCHG
    HASHSU
    s0 s4 s2 XC2PU
    CHKSIGNU
    34 THROWIFNOT
    ACCEPT
    s2 s0 XCPU
    SREFS
    IF:<{
      8 LDU
      LDREF
      s0 s2 XCHG
      SENDRAWMSG
    }>
    ENDS
    INC
    NEWC
    32 STU
    256 STU
    ENDC
    c4 POP
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      79,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      1,
      32,
      2,
      3,
      0,
      4,
      210,
      48,
      0,
      110,
      242,
      131,
      8,
      215,
      24,
      32,
      211,
      31,
      237,
      68,
      208,
      211,
      31,
      211,
      255,
      209,
      81,
      49,
      186,
      242,
      161,
      3,
      249,
      1,
      84,
      16,
      66,
      249,
      16,
      242,
      162,
      248,
      0,
      81,
      32,
      215,
      74,
      150,
      211,
      7,
      212,
      2,
      251,
      0,
      222,
      209,
      164,
      200,
      203,
      31,
      203,
      255,
      201,
      237,
      84,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2022.10 compiler 2`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC do_something
  DECLPROC do_another
  DECLPROC get_first
  DECLPROC get_second
  DECLPROC do_mutate
  DECLPROC main
  DECLGLOBVAR ctx
  do_something PROC:<{
    DUP
  }>
  do_another PROC:<{
    ADD
  }>
  get_first PROCINLINE:<{
    DROP
  }>
  get_second PROCINLINE:<{
    NIP
  }>
  do_mutate PROC:<{
    s2 s1 PUXC
    ADD
    -ROT
    ADD
  }>
  main PROC:<{
    1 PUSHINT
    do_something CALLDICT
    PAIR
    ctx SETGLOB
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      32,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      2,
      210,
      2,
      3,
      0,
      15,
      28,
      124,
      0,
      91,
      192,
      190,
      24,
      96,
      0,
      3,
      8,
      32,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2022.12 compiler 1`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC recv_internal
  DECLPROC recv_external
  recv_internal PROC:<{
    DROP
  }>
  recv_external PROC:<{
    9 PUSHPOW2
    LDSLICEX
    DUP
    32 LDU
    c4 PUSH
    CTOS
    32 LDU
    256 LDU
    ENDS
    s3 s1 XCPU
    EQUAL
    33 THROWIFNOT
    s0 s3 XCHG
    HASHSU
    s0 s4 s2 XC2PU
    CHKSIGNU
    34 THROWIFNOT
    ACCEPT
    s2 s0 XCPU
    SREFS
    IF:<{
      8 LDU
      LDREF
      s0 s2 XCHG
      SENDRAWMSG
    }>
    ENDS
    INC
    NEWC
    32 STU
    256 STU
    ENDC
    c4 POP
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      79,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      1,
      32,
      2,
      3,
      0,
      4,
      210,
      48,
      0,
      110,
      242,
      131,
      8,
      215,
      24,
      32,
      211,
      31,
      237,
      68,
      208,
      211,
      31,
      211,
      255,
      209,
      81,
      49,
      186,
      242,
      161,
      3,
      249,
      1,
      84,
      16,
      66,
      249,
      16,
      242,
      162,
      248,
      0,
      81,
      32,
      215,
      74,
      150,
      211,
      7,
      212,
      2,
      251,
      0,
      222,
      209,
      164,
      200,
      203,
      31,
      203,
      255,
      201,
      237,
      84,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2022.12 compiler 2`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC do_something
  DECLPROC do_another
  DECLPROC get_first
  DECLPROC get_second
  DECLPROC do_mutate
  DECLPROC main
  DECLGLOBVAR ctx
  do_something PROC:<{
    DUP
  }>
  do_another PROC:<{
    ADD
  }>
  get_first PROCINLINE:<{
    DROP
  }>
  get_second PROCINLINE:<{
    NIP
  }>
  do_mutate PROC:<{
    s2 s1 PUXC
    ADD
    -ROT
    ADD
  }>
  main PROC:<{
    1 PUSHINT
    do_something CALLDICT
    PAIR
    ctx SETGLOB
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      32,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      2,
      210,
      2,
      3,
      0,
      15,
      28,
      124,
      0,
      91,
      192,
      190,
      24,
      96,
      0,
      3,
      8,
      32,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2023.01 compiler 1`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC recv_internal
  DECLPROC recv_external
  recv_internal PROC:<{
    DROP
  }>
  recv_external PROC:<{
    9 PUSHPOW2
    LDSLICEX
    DUP
    32 LDU
    c4 PUSH
    CTOS
    32 LDU
    256 LDU
    ENDS
    s3 s1 XCPU
    EQUAL
    33 THROWIFNOT
    s0 s3 XCHG
    HASHSU
    s0 s4 s2 XC2PU
    CHKSIGNU
    34 THROWIFNOT
    ACCEPT
    s2 s0 XCPU
    SREFS
    IF:<{
      8 LDU
      LDREF
      s0 s2 XCHG
      SENDRAWMSG
    }>
    ENDS
    INC
    NEWC
    32 STU
    256 STU
    ENDC
    c4 POP
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      79,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      1,
      32,
      2,
      3,
      0,
      4,
      210,
      48,
      0,
      110,
      242,
      131,
      8,
      215,
      24,
      32,
      211,
      31,
      237,
      68,
      208,
      211,
      31,
      211,
      255,
      209,
      81,
      49,
      186,
      242,
      161,
      3,
      249,
      1,
      84,
      16,
      66,
      249,
      16,
      242,
      162,
      248,
      0,
      81,
      32,
      215,
      74,
      150,
      211,
      7,
      212,
      2,
      251,
      0,
      222,
      209,
      164,
      200,
      203,
      31,
      203,
      255,
      201,
      237,
      84,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2023.01 compiler 2`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC do_something
  DECLPROC do_another
  DECLPROC get_first
  DECLPROC get_second
  DECLPROC do_mutate
  DECLPROC main
  DECLGLOBVAR ctx
  do_something PROC:<{
    DUP
  }>
  do_another PROC:<{
    ADD
  }>
  get_first PROCINLINE:<{
    DROP
  }>
  get_second PROCINLINE:<{
    NIP
  }>
  do_mutate PROC:<{
    s2 s1 PUXC
    ADD
    -ROT
    ADD
  }>
  main PROC:<{
    1 PUSHINT
    do_something CALLDICT
    PAIR
    ctx SETGLOB
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      32,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      2,
      210,
      2,
      3,
      0,
      15,
      28,
      124,
      0,
      91,
      192,
      190,
      24,
      96,
      0,
      3,
      8,
      32,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2024.01 compiler 1`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC recv_internal
  DECLPROC recv_external
  recv_internal PROC:<{
    DROP
  }>
  recv_external PROC:<{
    9 PUSHPOW2
    LDSLICEX
    DUP
    32 LDU
    c4 PUSH
    CTOS
    32 LDU
    256 LDU
    ENDS
    s3 s1 XCPU
    EQUAL
    33 THROWIFNOT
    s0 s3 XCHG
    HASHSU
    s0 s4 s2 XC2PU
    CHKSIGNU
    34 THROWIFNOT
    ACCEPT
    s2 s0 XCPU
    SREFS
    IF:<{
      8 LDU
      LDREF
      s0 s2 XCHG
      SENDRAWMSG
    }>
    ENDS
    INC
    NEWC
    32 STU
    256 STU
    ENDC
    c4 POP
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      79,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      1,
      32,
      2,
      3,
      0,
      4,
      210,
      48,
      0,
      110,
      242,
      131,
      8,
      215,
      24,
      32,
      211,
      31,
      237,
      68,
      208,
      211,
      31,
      211,
      255,
      209,
      81,
      49,
      186,
      242,
      161,
      3,
      249,
      1,
      84,
      16,
      66,
      249,
      16,
      242,
      162,
      248,
      0,
      81,
      32,
      215,
      74,
      150,
      211,
      7,
      212,
      2,
      251,
      0,
      222,
      209,
      164,
      200,
      203,
      31,
      203,
      255,
      201,
      237,
      84,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile source with v2024.01 compiler 2`] = `
Object {
  "fift": "PROGRAM{
  DECLPROC do_something
  DECLPROC do_another
  DECLPROC get_first
  DECLPROC get_second
  DECLPROC do_mutate
  DECLPROC main
  DECLGLOBVAR ctx
  do_something PROC:<{
    DUP
  }>
  do_another PROC:<{
    ADD
  }>
  get_first PROCINLINE:<{
    DROP
  }>
  get_second PROCINLINE:<{
    NIP
  }>
  do_mutate PROC:<{
    s2 s1 PUXC
    ADD
    -ROT
    ADD
  }>
  main PROC:<{
    1 PUSHINT
    do_something CALLDICT
    PAIR
    ctx SETGLOB
  }>
}END>c
",
  "log": "",
  "ok": true,
  "output": Object {
    "data": Array [
      181,
      238,
      156,
      114,
      1,
      1,
      4,
      1,
      0,
      32,
      0,
      1,
      20,
      255,
      0,
      244,
      164,
      19,
      244,
      188,
      242,
      200,
      11,
      1,
      2,
      2,
      210,
      2,
      3,
      0,
      15,
      28,
      124,
      0,
      91,
      192,
      190,
      24,
      96,
      0,
      3,
      8,
      32,
    ],
    "type": "Buffer",
  },
}
`;

exports[`ton-compiler should compile with errors with legacy compiler 1`] = `
Object {
  "fift": null,
  "log": "src/tests/test_bug.fc:26:3: error: undefined function \`<some-syntax-error>\`, defining a global function of unknown type
    <some-syntax-error> 
    ^
src/tests/test_bug.fc:27:1: error: \`;\` expected instead of \`}\`
  } 
  ^

",
  "ok": false,
  "output": null,
}
`;

exports[`ton-compiler should compile with errors with v2022.10 compiler 1`] = `
Object {
  "fift": null,
  "log": "Func compilation error: /working/src/tests/test_bug.fc:27:1: error: \`;\` expected instead of \`}\`
  }",
  "ok": false,
  "output": null,
}
`;

exports[`ton-compiler should compile with errors with v2022.12 compiler 1`] = `
Object {
  "fift": null,
  "log": "Func compilation error: /working/src/tests/test_bug.fc:27:1: error: \`;\` expected instead of \`}\`
  }",
  "ok": false,
  "output": null,
}
`;

exports[`ton-compiler should compile with errors with v2023.01 compiler 1`] = `
Object {
  "fift": null,
  "log": "Func compilation error: /working/src/tests/test_bug.fc:27:1: error: \`;\` expected instead of \`}\`
  }",
  "ok": false,
  "output": null,
}
`;

exports[`ton-compiler should compile with errors with v2024.01 compiler 1`] = `
Object {
  "fift": null,
  "log": "Func compilation error: /working/src/tests/test_bug.fc:27:1: error: \`;\` expected instead of \`}\`
  }",
  "ok": false,
  "output": null,
}
`;
